lib/perl/Genome/Model/Tools/Speedseq/Base.pm

```diff
@@ -0,0 +1,290 @@
+    my $tool_param_name = $meta->tool_param_name;
+
+    return unless defined $self->$ur_name;
+
+    # This works for things that are is_many or not.
+    return map {_format_tool_arg($type, $tool_param_name, $_)} $self->$ur_name;
+}
+
+sub _tool_input_args_from_meta {
+    my ($self, $meta) = @_;
+
+    my $ur_name = $meta->property_name;
+    
+    return unless defined $self->$ur_name;
+
+    # This works for things that are is_many or not.
+    return map {$_} $self->$ur_name;
+}
+
+sub _tool_input_files_from_meta {
+    my ($self, $meta) = @_;
+
+    my $ur_name = $meta->property_name;
+    
+    return unless defined $self->$ur_name;
+
+    # This works for things that are is_many or not.
+    return map {$_} $self->$ur_name;
+}
+
+# command line args that come after the executable
+#
+# The default implementation just translates all the inputs that have
+# speedseq_param_name defined. Boolean values get translated to true/false,
+# and is many attributes get repeated, e.g.,
+#
+#  create_md5_file => {..., is => 'Boolean'}
+#       when set to something that evalutes to false
+#       yields
+#  qw(CREATE_MD5_FILE=false)
+#
+# and
+#
+#  program_name => {..., is_many => 1}
+#       when set to ['a', 'b', 'c']
+#       yields
+#  qw(PROGRAM_NAME=a PROGRAM_NAME=b PROGRAM_NAME=c)
+#
+
+sub _tool_param_args {
+    my $self = shift;
+
+    my @metas = $self->_tool_param_metas;
+    my @args = map {$self->_tool_param_args_from_meta($_)} @metas;
+
+    return @args;
+}
+
+sub _tool_input_args {
+    my $self = shift;
+
+    my @metas = $self->_tool_input_metas;
+    my @args = map {$self->_tool_input_args_from_meta($_)} @metas;
+
+    return @args;
+}
+
+# this is your chance to make a loud sound and die if a caller asks for
+# something ridiculous. params can also be edited here if that is somehow
+# appropriate.
+sub _validate_params {
+    # example:
+    # my $self = shift;
+    # $self->enforce_minimum_version('1.85');
+    # die unless -s $self->input_file;
+    # $self->be_noisy(0) unless $self->log_file;
+}
+
+# want to pass extra stuff to Genome::Sys->shellcmd? return a hash (not ref)
+# of those params here.
+sub _shellcmd_extra_params {
+    my $self = shift;
+    my @output_files = $self->_resolve_output_files;
+    my @input_files = $self->_resolve_input_files;
+    return (
+        skip_if_output_is_present => 0,
+        keep_dbh_connection_open => 0,
+        input_files => \@input_files,
+        output_files => \@output_files,
+    );
+}
+
+
+# string to be appended after the full command line (e.g., '> foo.out')
+#
+# there was a bug that led to some commands being built like:
+#
+#   java ... foo.jar class ARGS > output_file MORE_ARGS
+#
+# that's bad. this sort of conflicts with the idea of log_file, but commands
+# like "BamIndexStats" that have no native OUTPUT option (i.e., can only spew
+# to stdout) support the existence of this function.
+sub _redirects {
```

Comment:
I don't think this applies to speedseq - no need to keep everything from Picard that we don't need.